README.md
# Nerves.Leds

[![Build Status](https://travis-ci.org/nerves-project/nerves_leds.svg?branch=master)](https://travis-ci.org/nerves-project/nerves_leds)
[![Hex version](https://img.shields.io/hexpm/v/nerves_leds.svg "Hex version")](https://hex.pm/packages/nerves_leds)

Simple API to drive leds exposed by linux `/sys/class/leds`.  Designed for use with [Nerves](http://nerves-project.org/), but works on any distribution of Linux with `/sys/class/leds`.

## Installation

Add `nerves_leds` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:nerves_leds, "~> 0.8.1"}]
end
```

## Configuration & Usage

Configure your LEDs mappings:

```elixir
# config/config.exs
config :nerves_leds, names: [ error: "led0", connect: "led1" ]
```

Now, you can use predefined states in your app:

```elixir
alias Nerves.Leds

Leds.set(connect: true)
Leds.set(connect: :heartbeat)
Leds.set(connect: false, error: :slowblink)
```

You can also define your own states:

```elixir
# config/config.exs
config :nerves_leds, states: [
    fastblink: [ trigger: "timer", delay_off: 40, delay_on: 30 ],
    blip: [ trigger: "timer", delay_off: 1000, delay_on: 100 ]]
```

For situations where the number of LEDs are not known at compile time, you
can enumerate the LEDs:

```elixir
Leds.enumerate()
```

## More Details

See the [documentation](https://hexdocs.pm/nerves_leds) for the full description of the API and configuration.


CHANGELOG.md
# Changelog

## v0.8.1 (2020-10-01)

* Adds `enumerate` method for getting a list of available LEDs.

## v0.8.0 (2017-02-22)

* Specifying string for LED name to set function now bypasses name mapping
* Removed all compile-time config cacheing to avoid deps issues
* Now raises ArgumentError for invalid states
* Added alternate single-LED API form `set/2`
* Simplified README

## v0.7.0 (2016-06-21)

This update reflects a number of changes that bring this module more in line
with current Nerves and Elixir naming, coding, and license conventions, in
preparation for pushing to hex. Unfortunately it changes the API, but hey,
we're still < 1.0, right?

* Breaking API Changes
  - Renamed project to `nerves_leds` (was `nerves_io_led`)
  - Module is now Nerves.Leds (was Nerves.Led)
  - Removed the `alive` functionality (can use `oneshot` trigger instead)
* Improved code style in places and fixed warnings
* Improved and fixed bugs in docs
* Changed LICENSE to Apache 2

## v0.6.0 (2015-10-04)

* Breaking API Changes
  - app_name is now `nerves_io_led` for config.exs
  - changed configuration options to avoid confusing 'map' terminology
	- config option `name_map` is now just `names`
	- config option `state_map` is now just `states`

## v0.5.0 (2015-10-01)

* Move from Cellulose to Nerves Project
* Scope as Nerves.IO.Leds
* Add support for ExDoc
* Clean up README.md
* Add package support


config/test.exs
use Mix.Config

# uses a special callback form for an led that can be used to define special
# behavior for testing rather than using sys/class/leds

config :nerves_leds,
  names: [
    test_led: &Nerves.Leds.Test.on_led_write/1
  ]

# redefine led state map to be default, but with an additional custom
# state to handle test_states

config :nerves_leds,
  states: [
    true: [brightness: 1],
    false: [brightness: 0],
    slowblink: [trigger: "timer", delay_off: 250, delay_on: 250],
    fastblink: [trigger: "timer", delay_off: 80, delay_on: 50],
    slowwink: [trigger: "timer", delay_on: 1000, delay_off: 100],
    heartbeat: [trigger: "heartbeat"],
    test_state: [foo: 3, bar: 29, baz: %{}]
  ]


test/leds_test.exs
defmodule LedsTest do

  use ExUnit.Case
  
	# callback for the test led that simply sets keys in the process dictionary
  # where we can assert them.  This works because the LED module does
  # everything in process.
  def on_led_write({k, v}), do: Process.put(k, v)

  defp led(key), do: Process.get(key) 
 
  test "turning on an LED" do
		Leds.set test_led: true
	  assert led(:brightness) == "1"
		assert led(:trigger) == "none"
  end

  test "turn off an led" do
		Leds.set test_led: false
	  assert led(:brightness) == "0"
		assert led(:trigger) == "none"
  end

  test "test a complex state" do
		Leds.set test_led: :slowwink
		assert led(:trigger) == "timer"
		assert led(:delay_on) == "1000"
    assert led(:delay_off) == "100"
  end

  test "custom state map" do
		Leds.set test_led: :test_state
		assert led(:foo) == "3"
    assert led(:bar) == 29
    assert led(:baz) == %{}
  end

end


test/nerves_led_test.exs
defmodule Nerves.IO.LedTest do

  use ExUnit.Case
  alias Nerves.IO.Led

  # callback for the test led that simply sets keys in the process dictionary
  # where we can assert them.  This works because the LED module does
  # everything in process.
  def on_led_write({k, v}), do: Process.put(k, v)

  defp led(key), do: Process.get(key)

  test "turning on an LED" do
    Led.set test_led: true
    assert led(:brightness) == 1
    assert led(:trigger) == "none"
  end

  test "turn off an LED" do
    Led.set test_led: false
    assert led(:brightness) == 0
    assert led(:trigger) == "none"
  end

  test "test a complex state" do
    Led.set test_led: :slowwink
    assert led(:trigger) == "timer"
    assert led(:delay_on) == 1000
    assert led(:delay_off) == 100
  end

  test "custom state map" do
    Led.set test_led: :test_state
    assert led(:foo) == 3
    assert led(:bar) == 29
    assert led(:baz) == %{}
  end
end


mix.exs
defmodule Nerves.Leds.MixProject do
  use Mix.Project

  @version "0.8.1"
  @source_url "https://github.com/nerves-project/nerves_leds"

  def project do
    [
      app: :nerves_leds,
      version: @version,
      elixir: "~> 1.4",
      deps: deps(),
      package: package(),
      description: description(),
      docs: docs(),
      preferred_cli_env: %{
        docs: :docs,
        "hex.publish": :docs,
        "hex.build": :docs
      }
    ]
  end

  def application do
    []
  end

  defp description do
    "Functions to drive LEDs on embedded systems"
  end

  defp deps do
    [{:ex_doc, "~> 0.22", only: :docs, runtime: false}]
  end

  defp package do
    [
      files: ~w(lib config) ++ ~w(README.md CHANGELOG.md LICENSE mix.exs),
      licenses: ["MIT"],
      links: %{"GitHub" => @source_url}
    ]
  end

  defp docs do
    [
      main: "Nerves.Leds",
      extras: ["README.md", "CHANGELOG.md"],
      source_ref: "v#{@version}",
      source_url: @source_url
    ]
  end
end


lib/leds.ex
defmodule Leds do

  @moduledoc """
  Handles LED blinking/handling in a configurable way, providing an
  easy-to use interface to setting LEDs defined in /sys/class/leds:

		Leds.set power: true, alert: false, network: :fastblink

  """

  @sys_leds_path "/sys/class/leds/"

  @name_map  Application.get_env :leds, :name_map, []

  @state_map Application.get_env :leds, :state_map, [
		true:  [ brightness: "1" ],
    false: [ brightness: "0" ],
		slowblink: [ trigger: "timer", delay_off: "250", delay_on: "250" ],
		fastblink: [ trigger: "timer", delay_off: "80", delay_on: "50" ],
		slowwink:  [ trigger: "timer", delay_on: "1000", delay_off: "100" ],
		heartbeat: [ trigger: "heartbeat" ]
  ]

  def initialize do
    :ets.new :led_alive_processes, [:set, :public, :named_table]
  end

  defp led_path(led, attribute) do
    Path.join  @sys_leds_path <> "#{led}/", attribute
  end

  # if the value of a defined LED is a function, then call the function
  # with the key/value arguments   Primarly used for testing
  defp write(led_fn, {key, value}) when is_function(led_fn) do
		led_fn.({key, value})
  end
  defp write(led, {key, value}) when is_binary(led) do
    if is_atom(key), do: key = :erlang.atom_to_binary(key, :utf8)
    File.write(led_path(led, key), value)
  end

  defp set_raw_state(led, settings) do
    {trigger, settings} = Dict.pop settings, :trigger, "none"
    write(led, {:trigger, trigger})
    Enum.each settings, &(led |> write &1)
  end

  defp set_keyed_state({key, val}) do
    raw_led = Dict.get @name_map, key
    unless is_list(val), do: val = Dict.get @state_map, val
    set_raw_state raw_led, val
  end

  defp hold(led, ms) do
    set_keyed_state({led, true})
    :timer.sleep ms
    set_keyed_state({led, false})
  end

  @doc """
  Keeps an led on for the specified amount of time - 5 secs default

  Written in such a way that you can call alive from multiple processes,
  and they will overlap - i.e. each call resets the timer to another time
  period.  If the specified timeout is different, the last timout called
  is the timeout that is used.   When the timeout expires (with no other
  alive call for that led) the led is extinguished.

  The following example shows turning on an led labelled :activity.  The
  call must be executed every 2 seconds or more to keep the activity led
  lit:

      Leds.alive :activity, 2000

  WARNING: This is a moderate overhead function, and shouldn't be called
           every millisecond.  It's intended for longer intervals.  This
           could be fixed.  Pull requests welcomed.
  """
  def alive(led, ms \\ 5000) do
    pinger_pid = :erlang.spawn fn() -> hold(led, ms) end
    case :ets.lookup(:led_alive_processes, led) do
      [{^led, old_pid}] -> Process.exit(old_pid, :replaced)
      _ -> nil
    end
    :ets.insert :led_alive_processes, [{led, pinger_pid}]
  end

  @doc """
  Set led status.  Settings is an enumerable k/v.
  Like this: Leds.set power: true
  """
  def set(settings) do
    Enum.each settings, &(set_keyed_state(&1))
  end
end


lib/nerves_io_led.ex
defmodule Nerves.IO.Led do

  @moduledoc """
  Handles LED blinking/handling in a configurable way, providing an
  easy-to use interface to setting LEDs defined in `/sys/class/leds`.

  See README.md for overview information and configuration.
  """

  @app :nerves_io_led

  @predefined_states [
    true:  [ brightness: 1 ],
    false: [ brightness: 0 ],
    slowblink: [ trigger: "timer", delay_off: 250, delay_on: 250 ],
    fastblink: [ trigger: "timer", delay_off: 80, delay_on: 50 ],
    slowwink:  [ trigger: "timer", delay_on: 1000, delay_off: 100 ],
    heartbeat: [ trigger: "heartbeat" ]
  ]

  @sys_leds_path "/sys/class/leds/"

  @led_names  Application.get_env(@app, :names, [])
  @led_states Dict.merge(Application.get_env(@app, :states, []), @predefined_states)

  @doc "Must be called once (no parameters) at startup to setup associations"
  def initialize do
    :ets.new :led_alive_processes, [:set, :public, :named_table]
  end

  defp led_path(led, attribute) do
    Path.join  @sys_leds_path <> "#{led}/", attribute
  end

  # if the value of a defined LED is a function, then call the function
  # with the key/value arguments   Primarily used for testing
  defp write(led_fn, {key, value}) when is_function(led_fn) do
    led_fn.({key, value})
  end
  defp write(led, {key, value}) when is_binary(led) do
    if is_atom(key), do: key = :erlang.atom_to_binary(key, :utf8)
    File.write(led_path(led, key), to_string(value))
  end

  defp set_raw_state(led, settings) do
    {trigger, settings} = Dict.pop settings, :trigger, "none"
    write(led, {:trigger, trigger})
    Enum.each settings, &(led |> write(&1))
  end

  defp set_keyed_state({key, val}) do
    raw_led = case (Dict.get @led_names, key) do
      nil ->
        raise ArgumentError, """
        Attempt to set unknown led key. Check your led names in config.exs,
        or maybe you forgot to mix deps.compile leds after change?
        """
      led -> led
    end
    unless is_list(val), do: val = Dict.get @led_states, val
    set_raw_state raw_led, val
  end

  defp hold(led, ms) do
    set_keyed_state({led, true})
    :timer.sleep ms
    set_keyed_state({led, false})
  end

  @doc """
  Keeps an led on for the specified amount of time - 5 secs default

  Written in such a way that you can call alive from multiple processes,
  and they will overlap - i.e. each call resets the timer to another time
  period.  If the specified timeout is different, the last timeout called
  is the timeout that is used.   When the timeout expires (with no other
  alive call for that led) the led is extinguished.

  The following example shows turning on an led labeled :activity.  The
  call must be executed every 2 seconds or more to keep the activity led
  lit:

  ```
  alias Nerves.IO.Led
  ...
  Led.alive :activity, 2000
  ```

  WARNING: This is a moderate overhead function, and shouldn't be called
  every millisecond.  It's intended for longer intervals. Pull requests
  encouraged!
  """
  def alive(led, ms \\ 5000) do
    pinger_pid = :erlang.spawn fn() -> hold(led, ms) end
    case :ets.lookup(:led_alive_processes, led) do
      [{^led, old_pid}] -> Process.exit(old_pid, :replaced)
      _ -> nil
    end
    :ets.insert :led_alive_processes, [{led, pinger_pid}]
  end

  @doc """
  Set led status.  Settings is an enumerable k/v.  Like this:

  ```
  Led.set power: true
  ```
  """
  def set(settings) do
    Enum.each settings, &(set_keyed_state(&1))
  end

end


lib/nerves_io_leds.ex
defmodule Nerves.IO.Leds do

  @moduledoc """
  Handles LED blinking/handling in a configurable way, providing an
  easy-to use interface to setting LEDs defined in `/sys/class/leds`.

  See README.md for overview information and configuration.
  """

  @app :nerves_io_leds

  @predefined_states [
    true:  [ brightness: 1 ],
    false: [ brightness: 0 ],
    slowblink: [ trigger: "timer", delay_off: 250, delay_on: 250 ],
    fastblink: [ trigger: "timer", delay_off: 80, delay_on: 50 ],
    slowwink:  [ trigger: "timer", delay_on: 1000, delay_off: 100 ],
    heartbeat: [ trigger: "heartbeat" ]
  ]

  @sys_leds_path "/sys/class/leds/"

  @led_names  Application.get_env(@app, :names, [])
  @led_states Dict.merge(Application.get_env(@app, :states, []), @predefined_states)

  @doc "Must be called once (no parameters) at startup to setup associations"
  def initialize do
    :ets.new :led_alive_processes, [:set, :public, :named_table]
  end

  defp led_path(led, attribute) do
    Path.join  @sys_leds_path <> "#{led}/", attribute
  end

  # if the value of a defined LED is a function, then call the function
  # with the key/value arguments   Primarly used for testing
  defp write(led_fn, {key, value}) when is_function(led_fn) do
    led_fn.({key, value})
  end
  defp write(led, {key, value}) when is_binary(led) do
    if is_atom(key), do: key = :erlang.atom_to_binary(key, :utf8)
    File.write(led_path(led, key), to_string(value))
  end

  defp set_raw_state(led, settings) do
    {trigger, settings} = Dict.pop settings, :trigger, "none"
    write(led, {:trigger, trigger})
    Enum.each settings, &(led |> write &1)
  end

  defp set_keyed_state({key, val}) do
    raw_led = Dict.get @led_names, key
    unless is_list(val), do: val = Dict.get @led_states, val
    set_raw_state raw_led, val
  end

  defp hold(led, ms) do
    set_keyed_state({led, true})
    :timer.sleep ms
    set_keyed_state({led, false})
  end

  @doc """
  Keeps an led on for the specified amount of time - 5 secs default

  Written in such a way that you can call alive from multiple processes,
  and they will overlap - i.e. each call resets the timer to another time
  period.  If the specified timeout is different, the last timout called
  is the timeout that is used.   When the timeout expires (with no other
  alive call for that led) the led is extinguished.

  The following example shows turning on an led labelled :activity.  The
  call must be executed every 2 seconds or more to keep the activity led
  lit:

  ```
  alias Nerves.IO.Leds
  ...
  Leds.alive :activity, 2000
  ```

  WARNING: This is a moderate overhead function, and shouldn't be called
  every millisecond.  It's intended for longer intervals. Pull requests
  encouraged!
  """
  def alive(led, ms \\ 5000) do
    pinger_pid = :erlang.spawn fn() -> hold(led, ms) end
    case :ets.lookup(:led_alive_processes, led) do
      [{^led, old_pid}] -> Process.exit(old_pid, :replaced)
      _ -> nil
    end
    :ets.insert :led_alive_processes, [{led, pinger_pid}]
  end

  @doc """
  Set led status.  Settings is an enumerable k/v.  Like this:

  ```
  Leds.set power: true
  ```
  """
  def set(settings) do
    Enum.each settings, &(set_keyed_state(&1))
  end

end


lib/nerves_leds.ex
defmodule Nerves.Leds do
  @moduledoc """
  A convenient interface to setting LEDs defined in `/sys/class/leds`.

  While an application could write directly to /sys/class/leds, the main advantage
  of using this module is to provide a layer of abstraction that allows easily
  defining application-specific led names and states, like this:

  ```elixir
  alias Nerves.Leds

  Leds.set power: true                            # turn on the led we called "power"
  Leds.set power: :slowblink                      # make it blink slowly
  Leds.set connected: false, alert: :fastblink    # set multiple LED states at once

  ## alternate syntax via set/2

  Leds.set :power, :slowblink

  ```

  ## Configuration

  Use config.exs in your application to create a friendly name that maps to an
  entry in `/sys/class/leds` that make sense for your application.

  A trivial example for Raspberry Pi:

  ```elixir
  # in your app's config/config.exs:
  config :nerves_leds, names: [ red: "led0", green: "led1" ]
  ```

  A more useful example for the Alix 2D boards implementing a router:

  ```elixir
  # in your app's config/config.exs:
  config :nerves_leds, names: [
    power:     "alix:1",
    connected: "alix:2",
    alert:     "alix:3"
  ]
  ```

  ## Included states

  In addition to `true` (on) and `false` (off) the following atoms provide predefined
  behaviors:

  - `:slowblink` - turns on and off slowly (about twice a second)
  - `:fastblink` - turns on and off rapidly (about 7.5 times a second)
  - `:slowwink` - mostly on, but "winks off" once every second or so
  - `:heartbeat` - a heartbeat pattern

  ## Customizing states

  The standard LED states are defined as `@predefined_states` near the top of
  `lib/nerves_leds.ex`. You can change or add to them using config.exs as
  follows:

  ```elixir
  config :nerves_leds, states: [
  	fastblink: [ trigger: "timer", delay_off: 40, delay_on: 30 ],
  	blip: [ trigger: "timer", delay_off: 1000, delay_on: 100 ]]
  ```

  See the Linux documentation on `sys/class/leds` to understand the meaning of
  trigger, delay, brightness, and other settings.
  """

  @app :nerves_leds

  @predefined_states [
    true: [brightness: 1],
    false: [brightness: 0],
    slowblink: [trigger: "timer", delay_off: 250, delay_on: 250],
    fastblink: [trigger: "timer", delay_off: 80, delay_on: 50],
    slowwink: [trigger: "timer", delay_on: 1000, delay_off: 100],
    heartbeat: [trigger: "heartbeat"]
  ]

  @sys_leds_path "/sys/class/leds/"

  @doc """
  Enumerate all the LEDs on this device.
  If the device has no LEDs or doesn't support the Linux LED interface,
  an empty list is returned.
  """

  @spec enumerate() :: [binary]
  def enumerate() do
    case File.ls(@sys_leds_path) do
      {:ok, leds} -> leds
      _ -> []
    end
  end

  @doc """
  Set states of one or more LEDs by using their mapped name

  ~~~elixir
    Nerves.Leds.set power: true, error: :fastblink
  ~~~
  """

  @spec set(Keyword.t()) :: true
  def set(settings) do
    Enum.each(settings, fn {led, state} ->
      set(led, state)
    end)

    true
  end

  @doc """
  Set the state of a single LED

  ~~~elixir
  Nerves.Leds.set :power, true
  Nerves.Leds.set :backlight, brightness: 200
  ~~~

  Note that unlike set/1, this allows optionally directly naming the file of the led
  in /sys/class/leds by using a string name rather than an atom.

  ~~~elixir
  Nerves.Leds.set "led2", [ trigger: "timer", delay_on: 1000, delay_off: 1000 ]
  ~~~
  """

  @spec set(atom | binary, atom | Keyword.t()) :: :ok
  def set(led, state) do
    set_raw_state(raw_led(led), raw_state(state))
  end

  ### private ###

  defp led_path(led, attribute) do
    Path.join(@sys_leds_path, "#{led}/#{attribute}")
  end

  # if the value of a defined LED is a function, then call the function
  # with the key/value arguments Primarily used for testing
  defp write(led_fn, {key, value}) when is_function(led_fn) do
    led_fn.({key, value})
  end

  defp write(led, {key, value}) when is_binary(led) do
    File.write(led_path(led, key), to_string(value))
  end

  defp set_raw_state(led, settings) do
    {trigger, settings} = Keyword.pop(settings, :trigger, "none")
    write(led, {:trigger, trigger})
    Enum.each(settings, &write(led, &1))
  end

  # if parameter isn't a list, lookup state from state map or predefined states
  defp raw_state(val) when is_list(val), do: val
  defp raw_state(val) when is_integer(val), do: Integer.to_string(val)

  defp raw_state(val) when is_atom(val) do
    @app
    |> Application.get_env(:states, [])
    |> Keyword.merge(@predefined_states)
    |> Keyword.get(val)
    |> case do
      nil ->
        raise ArgumentError, "Attempt to set unknown LED state: #{inspect(val)}"

      state ->
        state
    end
  end

  # if parameter isn't a string, lookup led name from configured led name map
  defp raw_led(key) when is_binary(key), do: key

  defp raw_led(key) when is_atom(key) do
    @app
    |> Application.get_env(:names, [])
    |> Keyword.get(key)
    |> case do
      nil ->
        raise ArgumentError, """
        Attempt to set unknown LED key: #{inspect(key)}.
        Check your LED names in config.exs.
        """

      led ->
        led
    end
  end
end
